src/mappingObjectTypeToKotlinType.kt

```diff
@@ -61,19 +63,44 @@ class ObjectTypeToKotlinTypeMapperImpl(
         val cachedTraitName = cache[typeKey]
         if (cachedTraitName != null) return cachedTraitName
 
+        val usedTypeParams = translator.declarations.flatMap {
+            when (it) {
+                is Variable ->
+                    listOf(it.type.name)
+                is Function ->
+                    it.callSignature.params.map { it.type.name } + it.callSignature.returnType.name
+                else ->
+                    emptyList()
+            }
+        }.distinct()
+        val typeParams = typeParameterDeclarations.filter { usedTypeParams.contains(it.identifierName.text) }
+                .map { TypeParam(it.identifierName.text) }
+
         val traitName = "T$${n++}"
+        val traitNameWithTypeParams = traitName + typeParams.map { it.toString() }.join(",", startWithIfNotEmpty = "<", endWithIfNotEmpty = ">")
         translator.name = traitName
+        translator.typeParams = typeParams
 
         declarations.add(translator.result)
 
-        cache[typeKey] = traitName
-        return traitName
+        cache[typeKey] = traitNameWithTypeParams
+        return traitNameWithTypeParams
     }
 
     override fun getKotlinTypeForTypeAlias(typeAlias: String): TS.TypeNode? {
         return typeNodeByAlias.get(typeAlias)
     }
 
+    override fun <T> whileTranslatingSignature(signatureDeclaration: TS.SignatureDeclaration, function: () -> T): T {
+        val signatureTypeParams = signatureDeclaration.typeParameters?.arr ?: emptyArray()
+        try {
+            typeParameterDeclarations.addAll(signatureTypeParams)
```

Comment:
I was able to do as you asked with creating a new ObjectTypeToKotlinTypeMapper based on the current one.  I did add a